swatch-contracts/src/main/java/com/redhat/swatch/contract/service/SubscriptionSyncService.java

```diff
@@ -0,0 +1,522 @@
+import com.redhat.swatch.contract.product.umb.UmbSubscription;
+import com.redhat.swatch.contract.repository.BillingProvider;
+import com.redhat.swatch.contract.repository.OfferingEntity;
+import com.redhat.swatch.contract.repository.OfferingRepository;
+import com.redhat.swatch.contract.repository.SubscriptionEntity;
+import com.redhat.swatch.contract.repository.SubscriptionRepository;
+import com.redhat.swatch.contract.utils.SubscriptionDtoUtil;
+import io.micrometer.core.annotation.Timed;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.persistence.EntityManager;
+import jakarta.transaction.Transactional;
+import jakarta.ws.rs.BadRequestException;
+import java.time.OffsetDateTime;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.candlepin.clock.ApplicationClock;
+
+@ApplicationScoped
+@Slf4j
+@AllArgsConstructor
+public class SubscriptionSyncService {
+
+  private static final int SUBSCRIPTION_BATCH_SIZE = 1024;
+
+  private final CapacityReconciliationService capacityReconciliationService;
+  private final OfferingSyncService offeringSyncService;
+  private final SubscriptionService subscriptionService;
+  private final SubscriptionRepository subscriptionRepository;
+  private final OfferingRepository offeringRepository;
+  private final EntityManager entityManager;
+  private final ApplicationClock clock;
+  private final ApplicationConfiguration properties;
+  private final ProductDenylist productDenylist;
+  private final ObjectMapper objectMapper;
+
+  @Transactional
+  @Timed("swatch_subscription_reconcile_org")
+  public void reconcileSubscriptionsWithSubscriptionService(String orgId, boolean paygOnly) {
+    log.info("Syncing subscriptions for orgId={}", orgId);
+
+    var dtos = subscriptionService.getSubscriptionsByOrgId(orgId).stream();
+
+    // Filter out non PAYG subscriptions for faster processing when they are not needed.
+    // Slow processing was causing: https://issues.redhat.com/browse/ENT-5083
+
+    if (paygOnly) {
+      dtos =
+          dtos.filter(
+              subscription -> SubscriptionDtoUtil.extractBillingProviderId(subscription) != null);
+    }
+
+    var subIdToDtoMap =
+        dtos.filter(this::shouldSyncSub)
+            .collect(
+                Collectors.toMap(
+                    dto -> dto.getId().toString(),
+                    Function.identity(),
+                    (firstMatch, secondMatch) -> firstMatch));
+
+    List<SubscriptionEntity> subEntitiesForDeletion = new ArrayList<>();
+
+    Set<String> seenIds = new HashSet<>(subIdToDtoMap.keySet());
+
+    AtomicInteger subscriptionHandled = new AtomicInteger(0);
```

Comment:
Why did you switch from the batching implementation in SubscriptionSyncController? I would like to record the reasoning in the PR description for history.

Reply:
This batching implementation was used to flush/clear the current session by batches, and using a simpler counter was enough to do this. However, in order to keep the port logic as it was and GIT history more consistent, I also ported this utility and reverted it.